file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingFileSystemProvider.kt

```diff
@@ -0,0 +1,410 @@
+
+        // If we have both target and source caching paths we copy the attributes from source to target and run
+        // Files.move(source, target, *newOptions) during copyCachedAttributesTo
+        if (options.contains(StandardCopyOption.COPY_ATTRIBUTES) &&
+            source is FileAttributeCachingPath &&
+            target is FileAttributeCachingPath
+        ) {
+            // Filter out StandardCopyOption.COPY_ATTRIBUTES here because we dont want the moved file to repopulate
+            // the cache from the delegate provider/filesystem.
+            val newOptions = options.filter {
+                it != StandardCopyOption.COPY_ATTRIBUTES
+            }.toTypedArray()
+
+            source.copyCachedAttributesTo(target)
+            Files.move(delegateSourcePath, delegateTargetPath, *newOptions)
+        } else {
+            // If the StandardCopyOption.COPY_ATTRIBUTES option is not selected, there is no need to cache the
+            // attributes for the moved file.
+            Files.move(delegateSourcePath, delegateTargetPath, *options)
+        }
+    }
+
+    override fun isSameFile(path: Path, path2: Path): Boolean = Files.isSameFile(path, path2)
+
+    /**
+     * Tells whether a file is considered to be hidden. The exact definition of hidden is platform or provider
+     * dependent.
+     *
+     * On Windows a file is considered hidden if it isn't a directory and the DOS hidden attribute is set.
+     *
+     * On UNIX a file is considered to be hidden if its name begins with a period character ('.').
+     *
+     * Depending on the implementation, this method may access the [path] to determine if the file is
+     * considered hidden.
+     *
+     * @param path The path to check.
+     * @return `true` if the path is hidden, `false` otherwise.
+     * @throws IOException if an error occurs while accessing the underlying delegate provider.
+     */
+    @Throws(IOException::class)
+    override fun isHidden(path: Path): Boolean = if (
+        // Adapted from org/junit/jupiter/api/condition/OS.java to look up the operating system name
+        System.getProperty("os.name").lowercase(Locale.ENGLISH).contains("win")
+    ) {
+        val cachingPath = path.asCachingPath()
+        val attributesMap = cachingPath.getAllAttributesMatchingName("dos:*") ?: throw IOException(
+            "Could not get dos attributes from delegate filesystem."
+        )
+        val isHidden = attributesMap["dos:hidden"] as Boolean && !(attributesMap["directory"] as Boolean)
+        isHidden
+    } else {
+        val delegatePath = path.asCachingPath().delegate
+        val delegateProvider = delegatePath.fileSystem.provider()
+        delegateProvider.isHidden(delegatePath)
+    }
+
+    override fun getFileStore(path: Path): FileStore {
+        val delegatePath = path.asCachingPath().delegate
+        val providerDelegate = delegatePath.fileSystem.provider()
+        return providerDelegate.getFileStore(delegatePath)
```

Comment:
I'm a little puzzled as to _why_ we are doing stuff like this all of the time. It seems like we could just as easily call

```suggestion
        val provider = path.fileSystem.provider()
        return provider.getFileStore(path)
```

There are a lot of instances where I'm unsure of what the _benefit_ of casting truly is, as we don't currently _initialize_ path  caches on creation. Also, this begs the question: why even provide an override?

The same can be said for all of these methods:

`getFileStore`
`getFileAttributeView`
`newFileChannel`
`newAsynchronousFileChannel`
`createSymbolicLink`
`createLink`
`readSymbolicLink`

If we're going to override them, we should probably at least have some  test coverage.

Reply:
I cast to the file caching path to enforce using the delegate path/provider every time.

getFileAttributeView
getFileStore

These two functions I just want to ensure we are working with the delegate of any file caching path when getting the store/view and not referencing anything with the FileAttributeCachingFileSystem associated with the FileAttributeCachingFileSystemProvider itself. I could maybe remove their overrides, but this would be a small optimization.

newAsynchronousFileChannel
newByteChannel
newFileChannel
createSymbolicLink
createLink
readSymbolicLink

these functions throw unsupportedOperationException in the filesystemProvider base class, yet they are implemented in the filesystemProvider associated with the default filesystem (so thats why i override them and use/pass the delegate to them)